controllers/operatorpolicy_controller.go

```diff
@@ -108,167 +101,300 @@ func (r *OperatorPolicyReconciler) Reconcile(ctx context.Context, req ctrl.Reque
-	ctx context.Context,
+// buildOperatorGroup bootstraps the OperatorGroup spec defined in the operator policy
+// with the apiversion and kind in preparation for resource creation
+func buildOperatorGroup(
 	policy *policyv1beta1.OperatorPolicy,
-	cachedSubscription *unstructured.Unstructured,
-	cachedOGList []unstructured.Unstructured,
-) (ctrl.Result, error) {
-	//  Create og, then trigger reconcile
-	if len(cachedOGList) == 0 {
-		if strings.EqualFold(string(policy.Spec.RemediationAction), string(policyv1.Enforce)) {
-			ogSpec := buildOperatorGroup(policy)
-
-			err := r.Create(ctx, ogSpec)
-			if err != nil {
-				OpLog.Error(err, "Error while creating OperatorGroup")
-				r.setCompliance(ctx, policy, policyv1.NonCompliant)
+) (*operatorv1.OperatorGroup, error) {
+	operatorGroup := new(operatorv1.OperatorGroup)
 
-				return reconcile.Result{}, err
-			}
+	operatorGroup.Status.LastUpdated = &metav1.Time{} // without this, some conversions can panic
+	operatorGroup.SetGroupVersionKind(operatorGroupGVK)
 
-			// Created successfully, requeue result
-			r.setCompliance(ctx, policy, policyv1.NonCompliant)
+	sub := make(map[string]interface{})
 
-			return reconcile.Result{Requeue: true}, err
-		} else if policy.Spec.OperatorGroup != nil {
-			// If inform mode, keep going and return at the end without requeue. Before
-			// continuing, should check if required og spec is created to set compliance state
+	err := json.Unmarshal(policy.Spec.Subscription.Raw, &sub)
+	if err != nil {
+		return nil, fmt.Errorf("error unmarshalling subscription: %w", err)
+	}
 
-			// Set to non compliant because operatorgroup does not exist on cluster, but
-			// should still go to check subscription
-			r.setCompliance(ctx, policy, policyv1.NonCompliant)
-		}
+	subNamespace, ok := sub["namespace"].(string)
+	if !ok {
+		return nil, fmt.Errorf("namespace is required in spec.subscription")
 	}
 
-	// Create new subscription
-	if cachedSubscription == nil {
-		if strings.EqualFold(string(policy.Spec.RemediationAction), string(policyv1.Enforce)) {
-			subscriptionSpec := buildSubscription(policy)
+	// Create a default OperatorGroup if one wasn't specified in the policy
+	if policy.Spec.OperatorGroup == nil {
+		operatorGroup.ObjectMeta.SetNamespace(subNamespace)
+		operatorGroup.ObjectMeta.SetGenerateName(subNamespace + "-") // This matches what the console creates
+		operatorGroup.Spec.TargetNamespaces = []string{}
 
-			err := r.Create(ctx, subscriptionSpec)
-			if err != nil {
-				OpLog.Error(err, "Could not handle missing musthave object")
-				r.setCompliance(ctx, policy, policyv1.NonCompliant)
+		return operatorGroup, nil
+	}
 
-				return reconcile.Result{}, err
-			}
+	opGroup := make(map[string]interface{})
 
-			// Future work: Check availability/status of all resources
-			// managed by the OperatorPolicy before setting compliance state
-			r.setCompliance(ctx, policy, policyv1.Compliant)
+	err = json.Unmarshal(policy.Spec.OperatorGroup.Raw, &opGroup)
```

Comment:
Does this fail if you directly unmarshal it to the `operatorv1.OperatorGroupSpec`? If not, it'd allow you to skip a step.

Reply:
I see, thanks! I posted a comment on the types to continue this discussion.